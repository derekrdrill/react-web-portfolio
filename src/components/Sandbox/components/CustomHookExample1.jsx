import React from 'react';
import { useFetch } from '../../../hooks/useFetch';
import { List, ListItem, Typography } from '@mui/material';

export const CustomHookExample1Return = ({ loading, data }) =>
  loading ? (
    <Typography className='loading' component='h4' variant='h4'>
      Loading...
    </Typography>
  ) : (
    <List className='list'>
      {data.map(item => (
        <ListItem key={item.id}>{item.title}</ListItem>
      ))}
    </List>
  );

export const CustomHookExample1 = () => {
  const { data, loading } = useFetch('https://jsonplaceholder.typicode.com/posts', {});

  return <CustomHookExample1Return loading={loading} data={data} />;
};
