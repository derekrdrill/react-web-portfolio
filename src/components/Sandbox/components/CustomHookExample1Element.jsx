import React from 'react';
import PropTypes from 'prop-types';
import { List, ListItem, Typography } from '@mui/material';

export const CustomHookExample1Element = ({ loading, data }) =>
  loading ? (
    <Typography className='loading' component='h4' variant='h4'>
      Loading...
    </Typography>
  ) : (
    <List className='list'>
      {data.map(item => (
        <ListItem key={item.id}>{item.title}</ListItem>
      ))}
    </List>
  );

CustomHookExample1Element.propTypes = {
  loading: PropTypes.bool,
  data: PropTypes.object,
};
