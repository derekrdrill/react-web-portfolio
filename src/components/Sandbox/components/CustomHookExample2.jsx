import React from 'react';
import { useLocalStorage } from '../../../hooks/useLocalStorage';
import { Button, TextField } from '@mui/material';

export const CustomHookExample2 = () => {
  const [task, setTask] = useLocalStorage('task', '');
  const [tasks, setTasks] = useLocalStorage('tasks', []);
  const handleSetTask = e => setTask(e.currentTarget.value);
  const handleSubmit = () => {
    const taskObj = {
      task,
      commpleted: false,
      date: new Date().toLocaleDateString(),
    };

    setTasks([...tasks, taskObj]);
  };

  return (
    <>
      <div>
        <TextField label='Task' onChange={handleSetTask} value={task} />
      </div>
      <Button color='info' onClick={handleSubmit} variant='contained'>
        Submit
      </Button>
    </>
  );
};
