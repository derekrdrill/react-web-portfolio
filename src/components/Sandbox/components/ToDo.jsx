import React, { useEffect, useState, useRef } from 'react';
import { Typography } from '@mui/material';

export const ToDoTitle = ({ loading, toDo }) => (loading ? 'Loading...' : toDo.title);

export const ToDo = () => {
  const [loading, setLoading] = useState(false);
  const [toDo, setToDo] = useState({});

  const isMounted = useRef(true);

  useEffect(() => {
    fetch('https://jsonplaceholder.typicode.com/todos/1')
      .then(res => res.json())
      .then(data => {
        if (isMounted.current) {
          setTimeout(() => {
            setToDo(data);
            setLoading(false);
          }, 3000);
        }
      });

    return () => {
      isMounted.current = false;
    };
  }, [isMounted]);

  return (
    <Typography variant='h6' component='h6'>
      <ToDoTitle loading={loading} toDo={toDo} />
    </Typography>
  );
};
