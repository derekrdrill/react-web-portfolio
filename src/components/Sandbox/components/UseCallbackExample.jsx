import React, { useCallback, useState } from 'react';
import { Button, Typography } from '@mui/material';

export const UseCallbackExample = () => {
  const [tasks, setTasks] = useState([]);

  const addTask = useCallback(() => {
    setTasks(previousState => [...previousState, 'Some task']);
  }, [setTasks]);

  return (
    <>
      <AddTaskButton addTask={addTask} />
      {tasks.map((task, index) => (
        <Typography key={index} paragraph>
          {task}
        </Typography>
      ))}
    </>
  );
};

const AddTaskButton = React.memo(({ addTask }) => {
  console.log('button rendered');

  return (
    <Button variant='contained' onClick={addTask}>
      Add
    </Button>
  );
});
