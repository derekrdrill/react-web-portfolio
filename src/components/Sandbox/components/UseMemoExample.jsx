import React, { useEffect, useRef, useState, useMemo } from 'react';
import { Button, TextField, Typography } from '@mui/material';
import { SettingsInputComponent } from '@mui/icons-material';

export const UseMemoExample = () => {
  const [number, setNumber] = useState(1);
  const [count, setCount] = useState(0);
  const renders = useRef(1);

  const handleSetNumber = e => setNumber(e.currentTarget.value);

  const getSqRt = number => {
    for (let i = 0; i < 10000; i++) {
      console.log(i);
    }

    console.log('Expensive function called');
    return Math.sqrt(number);
  };

  const sqRt = useMemo(() => getSqRt(number), [number]);

  const handleReRenderClick = () => {
    setCount(prevCount => {
      console.log(prevCount + 1);
      return prevCount + 1;
    });
  };

  useEffect(() => {
    renders.current = renders.current + 1;
  });

  return (
    <div>
      <TextField label='Number' onChange={handleSetNumber} type='number' />
      <Typography paragraph>Renders: {renders.current}</Typography>
      <Typography paragraph>Square Root of Number is {sqRt}</Typography>
      <Button onClick={handleReRenderClick} variant='contained'>
        Re-render
      </Button>
    </div>
  );
};
