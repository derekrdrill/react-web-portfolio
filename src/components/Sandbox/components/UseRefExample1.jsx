import React, { useRef } from 'react';
import { Button, TextField } from '@mui/material';

export const UseRefExample1 = () => {
  const inputRef = useRef();

  const handleSubmit = e => {
    e.preventDefault();
    // console.log(inputRef.current.value);
  };

  return (
    <>
      <div>
        <TextField label='Input' placeholder='Type something...' inputRef={inputRef} />
      </div>
      <Button onClick={handleSubmit} variant='contained'>
        Submit
      </Button>
    </>
  );
};
