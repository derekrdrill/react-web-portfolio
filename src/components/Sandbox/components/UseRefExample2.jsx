import React, { useEffect, useRef, useState } from 'react';
import { TextField, Typography } from '@mui/material';

export const UseRefExample2 = () => {
  const [inputText, setInputText] = useState('');
  const renders = useRef(1);
  const prevText = useRef('');
  const handleSetInpuText = e => setInputText(e.currentTarget.value);

  useEffect(() => {
    renders.current = renders.current + 1;
    prevText.current = inputText;
  }, [inputText]);

  return (
    <>
      <Typography component='h2' variant='h5'>
        Renders: {renders.current}
      </Typography>
      <Typography component='h2' variant='h6'>
        Previous Text: {prevText.current}
      </Typography>
      <div>
        <TextField label='Input' onChange={handleSetInpuText} placeholder='Type something...' value={inputText} />
      </div>
    </>
  );
};
