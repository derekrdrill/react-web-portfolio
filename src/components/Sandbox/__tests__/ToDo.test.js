import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { ToDo, ToDoTitle } from '../components/ToDo';

const renderer = new ShallowRenderer();

describe('To Do tests', () => {
  it('renders correctly', () => {
    renderer.render(<ToDo />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('renders to do title', () => {
    const tests = [
      { loading: true, result: 'Loading...' },
      { loading: false, toDo: { title: 'Title' }, result: 'Title' },
    ];

    tests.forEach(({ loading, toDo, result }) => {
      renderer.render(<ToDoTitle loading={loading} toDo={toDo} />);
      const render = renderer.getRenderOutput();

      expect(render).toEqual(result);
    });
  });
});
