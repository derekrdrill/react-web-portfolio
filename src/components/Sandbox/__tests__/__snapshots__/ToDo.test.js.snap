// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To Do tests renders correctly 1`] = `
<ForwardRef(Typography)
  component="h6"
  variant="h6"
>
  <ToDoTitle
    loading={false}
    toDo={Object {}}
  />
</ForwardRef(Typography)>
`;
